File: apps/workspace/lib/learning-map-history.ts
import type { LearningNodeKind } from "./types.ts";

export type LearningMapHistoryChange =
  | "added"
  | "removed"
  | "node_changed"
  | "entered_frontier"
  | "left_frontier"
  | "relations_changed";

export interface LearningMapNodeHistorySnapshot {
  label: string;
  kind: LearningNodeKind;
  missionRelevance: "core" | "supporting" | "optional";
  frontier: boolean;
  relations: string[];
}

export interface LearningMapNodeHistoryEvent {
  revision: number;
  updatedAt: string;
  rationale: string;
  evidenceCount: number;
  changes: LearningMapHistoryChange[];
  before?: LearningMapNodeHistorySnapshot;
  after?: LearningMapNodeHistorySnapshot;
}

export interface LearningMapNodeHistory {
  projectId: string;
  nodeId: string;
  currentRevision: number;
  events: LearningMapNodeHistoryEvent[];
}

const CHANGE_KINDS = new Set<LearningMapHistoryChange>([
  "added",
  "removed",
  "node_changed",
  "entered_frontier",
  "left_frontier",
  "relations_changed",
]);
const NODE_KINDS = new Set<LearningNodeKind>(["concept", "procedure", "strategy"]);
const RELEVANCE = new Set(["core", "supporting", "optional"] as const);

function object(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${label} must be an object`);
  }
  return value as Record<string, unknown>;
}

function text(value: unknown, label: string): string {
  if (typeof value !== "string" || !value.trim()) throw new Error(`${label} is invalid`);
  return value.trim();
}

function integer(value: unknown, label: string): number {
  if (!Number.isInteger(value) || (value as number) < 0) throw new Error(`${label} is invalid`);
  return value as number;
}

function strings(value: unknown, label: string): string[] {
  if (!Array.isArray(value) || value.some((item) => typeof item !== "string" || !item.trim())) {
    throw new Error(`${label} is invalid`);
  }
  return value.map((item) => (item as string).trim());
}

function snapshot(value: unknown, label: string): LearningMapNodeHistorySnapshot | undefined {
  if (value === null) return undefined;
  const item = object(value, label);
  const kind = text(item.kind, `${label}.kind`) as LearningNodeKind;
  const missionRelevance = text(item.mission_relevance, `${label}.mission_relevance`) as "core" | "supporting" | "optional";
  if (!NODE_KINDS.has(kind)) throw new Error(`${label}.kind is invalid`);
  if (!RELEVANCE.has(missionRelevance)) throw new Error(`${label}.mission_relevance is invalid`);
  if (typeof item.frontier !== "boolean") throw new Error(`${label}.frontier is invalid`);
  return {
    label: text(item.label, `${label}.label`),
    kind,
    missionRelevance,
    frontier: item.frontier,
    relations: strings(item.relations, `${label}.relations`),
  };
}

export function parseLearningMapNodeHistory(
  value: unknown,
  expectedProjectId: string,
  expectedNodeId: string,
): LearningMapNodeHistory {
  const root = object(value, "LearningMap node history");
  const projectId = text(root.project_id, "history project_id");
  const nodeId = text(root.node_id, "history node_id");
  if (projectId !== expectedProjectId) throw new Error("LearningMap history Project scope is invalid");
  if (nodeId !== expectedNodeId) throw new Error("LearningMap history node scope is invalid");
  if (!Array.isArray(root.events)) throw new Error("LearningMap history events are invalid");
  const currentRevision = integer(root.current_revision, "history current_revision");
  let previousRevision = Number.POSITIVE_INFINITY;
  const events = root.events.map((raw, index): LearningMapNodeHistoryEvent => {
    const item = object(raw, `history event ${index}`);
    const revision = integer(item.revision, `history event ${index}.revision`);
    if (revision > currentRevision || revision >= previousRevision) {
      throw new Error("LearningMap history event ordering is invalid");
    }
    previousRevision = revision;
    const changes = strings(item.changes, `history event ${index}.changes`) as LearningMapHistoryChange[];
    if (!changes.length || changes.some((change) => !CHANGE_KINDS.has(change))) {
      throw new Error("LearningMap history change kinds are invalid");
    }
    const before = snapshot(item.before, `history event ${index}.before`);
    const after = snapshot(item.after, `history event ${index}.after`);
    if (!before && !after) throw new Error("LearningMap history event must have before or after state");
    return {
      revision,
      updatedAt: text(item.updated_at, `history event ${index}.updated_at`),
      rationale: text(item.rationale, `history event ${index}.rationale`),
      evidenceCount: integer(item.evidence_count, `history event ${index}.evidence_count`),
      changes,
      before,
      after,
    };
  });
  return { projectId, nodeId, currentRevision, events };
}

File: tools/learning_map_history.py
#!/usr/bin/env python3
"""Read-only node history over validated append-only LearningMap revisions."""

from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any

try:
    from tools import learning_map, project_store
except ImportError:  # Direct execution
    import learning_map
    import project_store


class LearningMapHistoryError(RuntimeError):
    pass


def _resolve_context(repo_root: Path) -> project_store.ProjectContext:
    context = project_store.resolve_project_context(repo_root.resolve())
    if context.layout != project_store.LAYOUT_WORKSPACE or context.learning_map_path is None:
        raise LearningMapHistoryError("LearningMap history requires a selected workspace-v0.2 Project")
    return context


def _node_id(value: str) -> str:
    try:
        return project_store.validate_local_id(value, "node_id")
    except project_store.ProjectStoreError as exc:
        raise LearningMapHistoryError(str(exc)) from exc


def _load_revisions(
    repo_root: Path,
    context: project_store.ProjectContext,
) -> list[dict[str, Any]]:
    assert context.learning_map_path is not None
    root = context.learning_map_path.parent / "revisions"
    if root.is_symlink():
        raise LearningMapHistoryError("LearningMap revisions directory must not be a symbolic link")
    if not root.is_dir():
        return []
    revisions: list[dict[str, Any]] = []
    for path in sorted(root.glob("*.json")):
        try:
            value = learning_map.validate_learning_map(
                learning_map._read_object(path, "LearningMap revision"),
                context.project_id,
            )
        except learning_map.LearningMapError as exc:
            raise LearningMapHistoryError(str(exc)) from exc
        revisions.append(value)
    for index, revision in enumerate(revisions):
        if revision["revision"] != index:
            raise LearningMapHistoryError("LearningMap revision history is not contiguous")
        expected_parent = None if index == 0 else index - 1
        if revision["parent_revision"] != expected_parent:
            raise LearningMapHistoryError("LearningMap revision parent chain is invalid")
    current = learning_map.load_learning_map(repo_root.resolve())
    if current is not None and (not revisions or revisions[-1] != current):
        raise LearningMapHistoryError("LearningMap current revision does not match immutable history")
    return revisions


def _relations(value: dict[str, Any], node_id: str) -> list[str]:
    labels = {item["id"]: item["label"] for item in value["nodes"]}
    result: list[str] = []
    for edge in value["edges"]:
        if edge["source"] == node_id:
            target = labels.get(edge["target"], edge["target"])
            result.append(f"out: {edge['relation']} → {target} ({edge['confidence']})")
        elif edge["target"] == node_id:
            source = labels.get(edge["source"], edge["source"])
            result.append(f"in: {source} → {edge['relation']} ({edge['confidence']})")
    return sorted(result)


def _snapshot(value: dict[str, Any], node_id: str) -> dict[str, Any] | None:
    node = next((item for item in value["nodes"] if item["id"] == node_id), None)
    if node is None:
        return None
    return {
        "label": node["label"],
        "kind": node["kind"],
        "mission_relevance": node["mission_relevance"],
        "frontier": node_id in value["frontier"],
        "relations": _relations(value, node_id),
    }


def _changes(before: dict[str, Any] | None, after: dict[str, Any] | None) -> list[str]:
    if before is None and after is None:
        return []
    if before is None:
        result = ["added"]
        if after and after["frontier"]:
            result.append("entered_frontier")
        if after and after["relations"]:
            result.append("relations_changed")
        return result
    if after is None:
        result = ["removed"]
        if before["frontier"]:
            result.append("left_frontier")
        if before["relations"]:
            result.append("relations_changed")
        return result

    result: list[str] = []
    if any(before[field] != after[field] for field in ("label", "kind", "mission_relevance")):
        result.append("node_changed")
    if before["frontier"] != after["frontier"]:
        result.append("entered_frontier" if after["frontier"] else "left_frontier")
    if before["relations"] != after["relations"]:
        result.append("relations_changed")
    return result


def node_history(repo_root: Path, node_id: str, *, limit: int = 20) -> dict[str, Any]:
    repo_root = repo_root.resolve()
    context = _resolve_context(repo_root)
    node_id = _node_id(node_id)
    if isinstance(limit, bool) or not isinstance(limit, int) or limit < 1 or limit > 100:
        raise LearningMapHistoryError("limit must be an integer from 1 to 100")
    revisions = _load_revisions(repo_root, context)
    if not revisions:
        return {
            "project_id": context.project_id,
            "node_id": node_id,
            "current_revision": 0,
            "events": [],
        }

    events: list[dict[str, Any]] = []
    previous: dict[str, Any] | None = None
    for revision in revisions:
        after = _snapshot(revision, node_id)
        change_kinds = _changes(previous, after)
        if change_kinds:
            events.append({
                "revision": revision["revision"],
                "updated_at": revision["updated_at"],
                "rationale": revision["rationale"],
                "evidence_count": len(revision["evidence_ids"]),
                "changes": change_kinds,
                "before": previous,
                "after": after,
            })
        previous = after

    return {
        "project_id": context.project_id,
        "node_id": node_id,
        "current_revision": revisions[-1]["revision"],
        "events": list(reversed(events[-limit:])),
    }


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="ai4learning LearningMap node history reader")
    parser.add_argument("--repo", type=Path, default=Path.cwd(), help="repository root")
    sub = parser.add_subparsers(dest="command", required=True)
    node = sub.add_parser("node", help="show meaningful revision events for one node")
    node.add_argument("node_id")
    node.add_argument("--limit", type=int, default=20)
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    try:
        value = node_history(args.repo.resolve(), args.node_id, limit=args.limit)
        print(json.dumps(value, ensure_ascii=False, indent=2))
        return 0
    except (LearningMapHistoryError, project_store.ProjectStoreError) as exc:
        print(f"error: {exc}", file=__import__("sys").stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())
